Model/PageType.php

```diff
@@ -0,0 +1,223 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Magento\UpwardConnector\Model;
+
+use Magento\Framework\Serialize\Serializer\Json;
+use Magento\Store\Model\StoreManagerInterface;
+use Magento\UrlRewrite\Model\UrlFinderInterface;
+use Magento\UrlRewrite\Service\V1\Data\UrlRewrite;
+use Magento\UrlRewriteGraphQl\Model\Resolver\UrlRewrite\CustomUrlLocatorInterface;
+use Magento\Upward\Context;
+
+class PageType
+{
+    /** @var \Magento\Store\Model\StoreManagerInterface */
+    private $storeManager;
+
+    /** @var \Magento\UrlRewrite\Model\UrlFinderInterface */
+    private $urlFinder;
+
+    /** @var \Magento\UrlRewriteGraphQl\Model\Resolver\UrlRewrite\CustomUrlLocatorInterface */
+    private $customUrlLocator;
+
+    /** @var \Magento\Framework\Serialize\Serializer\Json */
+    private $json;
+
+    /** @var int */
+    private $redirectType;
+
+    /** @var \Magento\Upward\Context */
+    private $context;
+
+    /** @var array|null */
+    private $pageType;
+
+    public function __construct(
+        StoreManagerInterface $storeManager,
+        UrlFinderInterface $urlFinder,
+        CustomUrlLocatorInterface $customUrlLocator,
+        Json $json
+    ) {
+        $this->storeManager = $storeManager;
+        $this->urlFinder = $urlFinder;
+        $this->customUrlLocator = $customUrlLocator;
+        $this->json = $json;
+    }
+
+    public function getInfo(): ?array
+    {
+        if ($this->pageType === null) {
+            $this->pageType = $this->resolvePageInfo();
+        }
+
+        return $this->pageType;
+    }
+
+    public function getJson(): string
+    {
+        $pageType = $this->getInfo();
+
+        return $pageType ? $this->json->serialize($pageType) : '';
+    }
+
+    public function getPageType(): ?string
+    {
+        $pageInfo = $this->getInfo();
+
+        return $pageInfo['type'] ?? null;
+    }
+
+    public function setContext(Context $context): self
+    {
+        $this->context = $context;
+
+        return $this;
+    }
+
+    public function getContext(): ?Context
+    {
+        return $this->context;
+    }
+
+    public function resolvePageInfo(): ?array
+    {
+        $storeId = (int) $this->storeManager->getStore()->getId();
+        $request = $this->getContext()->get('request')->toArray();
+        $urlParts = $request['url'];
+        $url = $urlParts['pathname'];
+        $result = null;
+
+        if (substr($url, 0, 1) === '/' && $url !== '/') {
+            $url = ltrim($url, '/');
+        }
+        $this->redirectType = 0;
+        $customUrl = $this->customUrlLocator->locateUrl($url);
+        $url = $customUrl ?: $url;
+        $finalUrlRewrite = $this->findFinalUrl($url, $storeId);
+        if ($finalUrlRewrite) {
```

Comment:
```
  if ($finalUrlRewrite) {
            return null;
        }

        $relativeUrl = $finalUrlRewrite->getRequestPath();
        $resultArray = $this->rewriteCustomUrls($finalUrlRewrite, $storeId) ?? [
                'id' => $finalUrlRewrite->getEntityId(),
                'canonical_url' => $relativeUrl,
                'relative_url' => $relativeUrl,
                'redirectCode' => $this->redirectType,
                'redirect_code' => $this->redirectType,
                'type' => $this->sanitizeType($finalUrlRewrite->getEntityType())
            ];
        if (!empty($urlParts['search'])) {
            $resultArray['relative_url'] .= $urlParts['search'];
        }

        if (empty($resultArray['id'])) {
            throw new \RuntimeException(
                __('No such entity found with matching URL key: %url', ['url' => $url])
            );
        }

       return $resultArray ?: null;
```

Reply:
Since this chunk is 99% from core (UrlRewriteGraphQl), I'd prefer to keep the structure as is, to make updating easier if ever core is changed